sputnikdao-factory2/src/factory_manager.rs

```diff
@@ -0,0 +1,160 @@
+    pub fn store_contract(&self) {
+        unsafe {
+            // Load input into register 0.
+            sys::input(0);
+            // Compute sha256 hash of register 0 and store in register 1.
+            sys::sha256(u64::MAX as _, 0 as _, 1);
+            // Check if such blob is already stored.
+            assert_eq!(
+                sys::storage_has_key(u64::MAX as _, 1 as _),
+                0,
+                "ERR_ALREADY_EXISTS"
+            );
+            // Store key-value pair. The key is represented by register 1 and the value by register 0.
+            sys::storage_write(u64::MAX as _, 1 as _, u64::MAX as _, 0 as _, 2);
+            // Load register 1 into blob_hash.
+            let blob_hash = [0u8; 32];
+            sys::read_register(1, blob_hash.as_ptr() as _);
+            // Return from function value of register 1.
+            let blob_hash_str = serde_json::to_string(&Base58CryptoHash::from(blob_hash))
+                .unwrap()
+                .into_bytes();
+            sys::value_return(blob_hash_str.len() as _, blob_hash_str.as_ptr() as _);
+        }
+    }
+
+    /// Delete code from the contract.
+    pub fn delete_contract(&self, code_hash: Base58CryptoHash) {
+        let code_hash: CryptoHash = code_hash.into();
+        env::storage_remove(&code_hash);
+    }
+
+    /// Get code for given hash.
+    pub fn get_code(&self, code_hash: Base58CryptoHash) {
+        let code_hash: CryptoHash = code_hash.into();
+        unsafe {
+            // Check that such contract exists.
+            assert!(env::storage_has_key(&code_hash), "Contract doesn't exist");
+            // Load the hash from storage.
+            sys::storage_read(code_hash.len() as _, code_hash.as_ptr() as _, 0);
+            // Return as value.
+            sys::value_return(u64::MAX as _, 0 as _);
+        }
+    }
+
+    /// Forces update on the given contract.
+    /// Contract must support update by factory for this via permission check.
+    pub fn update_contract(
+        &self,
+        account_id: AccountId,
+        code_hash: Base58CryptoHash,
+        method_name: &str,
+    ) {
+        let code_hash: CryptoHash = code_hash.into();
+        let account_id = account_id.as_bytes().to_vec();
+        unsafe {
+            // Check that such contract exists.
+            assert!(env::storage_has_key(&code_hash), "Contract doesn't exist");
+            // Load the hash from storage.
+            sys::storage_read(code_hash.len() as _, code_hash.as_ptr() as _, 0);
+            // Create a promise toward given account.
+            let promise_id =
+                sys::promise_batch_create(account_id.len() as _, account_id.as_ptr() as _);
+            // Call `update` method, which should also handle migrations.
+            sys::promise_batch_action_function_call(
+                promise_id,
+                method_name.len() as _,
+                method_name.as_ptr() as _,
+                u64::MAX as _,
+                0,
```

Comment:
I think what you actually need here is not 0, but `sys::value_return(u64::MAX as _, 0 as _);`. After this, on the DAO side, you can extract the value with `sys::input`.